### PR TITLE
escape bookmark and remote names in `trunk()` config

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/revset.rs [hunk 1]
```diff
@@ -2704,6 +2704,21 @@ pub struct RevsetWorkspaceContext<'a> {
     pub workspace_id: &'a WorkspaceId,
 }
 
+/// Formats a string as symbol by quoting and escaping it if necessary
+pub fn format_symbol(literal: &str) -> String {
+    let needs_escaping = if let Ok(node) = revset_parser::parse_program(literal) {
+        !matches!(node.kind, ExpressionKind::Identifier(_))
```
Comment (yuja) [inline lib/src/revset.rs:2710]: This doesn't work if the literal contains parentheses (e.g. `(foo)`.) We'll need to compare `node.span`.

It might be also good to add a test function by using low-level pest API, something like `RevsetParser::parse(Rule::identifier, s)` and compare span.

File: lib/src/revset.rs [hunk 2]
```diff
@@ -2704,6 +2704,21 @@ pub struct RevsetWorkspaceContext<'a> {
...
+    } else {
+        true
+    };
+    if needs_escaping {
+        let escaped = literal.replace('\\', r#"\\"#).replace('"', r#"\""#);
```
Comment (yuja) [inline lib/src/revset.rs:2715]: Maybe we can apply `char::escape_default()` to chars lower than `0x80`?
https://doc.rust-lang.org/stable/std/primitive.char.html#method.escape_default

`str::escape_debug()` might also work, but it might escape some unicode characters as `\u....` (which our parser doesn't support.)

btw, I would add `dsl_util::escape_string()` where string parser is defined.

File: lib/src/revset.rs (hunk 1 above)
Comment (yuja) [reply lib/src/revset.rs]: > Good point about those silent parsing rules. I update the code to compare the span. Just for my understanding, would it be equivalent to compare the parsed identifier (in `ExpressionKind::Identifier(identifier)`)?

I think it'll work right now (because identifier doesn't support escape syntax.) It might break in future if we change the grammar.

> > It might be also good to add a test function by using low-level pest API, something like `RevsetParser::parse(Rule::identifier, s)` and compare span.
> 
> I didn't follow this part.

I just meant something like this would work (because we don't have to parse the full revset expression):
```rust
fn is_identifier(text: &str) -> bool {
    match RevsetParser::parse(Rule::identifier, text) {
        Ok(pairs) => pairs.next().unwrap().as_span().end() == text.len(),
        Err(_) => false,
    }
}
```
(untested)

File: lib/src/revset.rs (hunk 2 above)
Comment (yuja) [reply lib/src/revset.rs]: > > Maybe we can apply `char::escape_default()` to chars lower than `0x80`? https://doc.rust-lang.org/stable/std/primitive.char.html#method.escape_default
> 
> Why? Would it prevent some bugs? Or would it be to make the stored string not rely on unicode in order to make sure it's readable even if you have a simple terminal or font?

Just because it would be easier to implement. Suppose we want to reuse `escape_string()` for displaying purpose, I expect control characters in ASCII range are escaped.

Comment (yuja) [reply lib/src/revset.rs]: I assumed `char::escape_default()` would use '\xNN` for lower values, but apparently not. It is the behavior of `ascii::escape_default()`.

File: lib/src/revset.rs (hunk 1 above)
Comment (martinvonz) [reply lib/src/revset.rs]: Ah, makes sense. That's much cleaner. Done.

File: lib/src/revset.rs (hunk 2 above)
Comment (martinvonz) [reply lib/src/revset.rs]: I think it's nice if we can allow non-ascii values without escaping, so bookmark names like the one in #5359 are easy to read.

Comment (yuja) [reply lib/src/revset.rs]: > we can allow non-ascii values without escaping,

Yes, I agree. I just mean it's better to escape ASCII control characters (i.e. 0x00..=0x1f and 0x7f.)

File: cli/tests/test_git_clone.rs [hunk 3]
```diff
@@ -599,6 +599,36 @@ fn test_git_clone_remote_default_bookmark(subprocess: bool) {
...
+        insta::assert_snapshot!(stderr, @r#"
+    Fetching into new repo in "$TEST_ENV/clone4"
+    bookmark: "@origin        [new] untracked
+    bookmark: feature1@origin [new] untracked
+    bookmark: main@origin     [new] untracked
+    Setting the revset alias "trunk()" to ""\""@origin"
+    Working copy now at: wmwvqwsz abb7c50b (empty) (no description set)
+    Parent commit      : mzyxwzks 9f01a0e0 " feature1@origin main@origin | message
+    Added 1 files, modified 0 files, removed 0 files
+    "#);
```
Comment (yuja) [inline cli/tests/test_git_clone.rs:620]: nit: rerun `insta` to update the indent (or split this to separate test function)?

File: lib/src/dsl_util.rs [hunk 4]
```diff
@@ -429,6 +429,24 @@ impl<R: RuleType> StringLiteralParser<R> {
...
+    let mut escaped = String::with_capacity(unescaped.len());
+    for c in unescaped.chars() {
+        match c {
+            '"' => escaped.push_str(r#"\""#),
+            '\\' => escaped.push_str(r#"\\"#),
+            '\t' => escaped.push_str(r#"\t"#),
+            '\r' => escaped.push_str(r#"\r"#),
+            '\n' => escaped.push_str(r#"\n"#),
+            '\0' => escaped.push_str(r#"\0"#),
+            '\x1b' => escaped.push_str(r#"\e"#),
```
Comment (yuja) [inline lib/src/dsl_util.rs:443]: nit: I would escape `\x1b` in hex form since `\e` isn't common escape sequence.

File: lib/src/revset.rs [hunk 5]
```diff
@@ -4327,4 +4336,29 @@ mod tests {
...
+        assert_eq!(format_symbol(r#"foo\bar"#), r#""foo\\bar""#);
+        assert_eq!(format_symbol(r#"foo\"bar"#), r#""foo\\\"bar""#);
+        assert_eq!(format_symbol("foo\nbar"), r#""foo\nbar""#);
+
+        // Some characters don't technically need escaping, but we escape them for
+        // clarity
+        assert_eq!(format_symbol(r#"foo"bar"#), r#""foo\"bar""#);
+        assert_eq!(format_symbol(r#"foo\bar"#), r#""foo\\bar""#);
+        assert_eq!(format_symbol(r#"foo\"bar"#), r#""foo\\\"bar""#);
+    }
```
Comment (yuja) [inline lib/src/revset.rs:4363]: nit: can you add test for symbol starts/ends with space? It could be an edge case if our grammar had implicit whitespace rule.

**Review by yuja — APPROVED** [review]
LG, thanks.

File: cli/tests/test_git_clone.rs (hunk 3 above)
Comment (martinvonz) [reply cli/tests/test_git_clone.rs]: Good idea to split it into a separate function.

File: lib/src/revset.rs (hunk 2 above)
Comment (martinvonz) [reply lib/src/revset.rs]: Oh, I see. Yes, I agree. I'll do that.

Comment (martinvonz) [reply lib/src/revset.rs]: Done now.

File: lib/src/dsl_util.rs [hunk 6]
```diff
@@ -429,6 +430,28 @@ impl<R: RuleType> StringLiteralParser<R> {
...
+    let mut escaped = String::with_capacity(unescaped.len());
+    for c in unescaped.chars() {
+        match c {
+            '"' => escaped.push_str(r#"\""#),
+            '\\' => escaped.push_str(r#"\\"#),
+            '\t' => escaped.push_str(r#"\t"#),
+            '\r' => escaped.push_str(r#"\r"#),
+            '\n' => escaped.push_str(r#"\n"#),
+            '\0' => escaped.push_str(r#"\0"#),
+            c if c < '\x20' || c == '\x7f' => {
```
Comment (yuja) [inline lib/src/dsl_util.rs:444]: nit: `.is_ascii_contro()` or range pattern can be used

File: lib/src/dsl_util.rs [hunk 7]
```diff
@@ -14,6 +14,7 @@
 
 //! Domain-specific language helpers.
 
+use core::ascii;
```
Comment (yuja) [inline lib/src/dsl_util.rs:17]: nit: just import from `std::`?

**Review by yuja — APPROVED** [review]

File: lib/src/dsl_util.rs (hunk 6 above)
Comment (martinvonz) [reply lib/src/dsl_util.rs:444]: Ah, nice. Switched to `.is_ascii_control()`

File: lib/src/dsl_util.rs (hunk 7 above)
Comment (martinvonz) [reply lib/src/dsl_util.rs:17]: Done. (VS code does the imports and I usually don't check.)
